### PR TITLE
translate std_misc/threads

### LINKED ISSUE #90: std_misc/threads
1. [testcase_mapreduce.md](https://github.com/nebula-labs/rust-by-example-vn/blob/main/src/std_misc/threads/testcase_mapreduce.md)
2. [arg.md](https://github.com/nebula-labs/rust-by-example-vn/blob/main/src/std_misc/arg.md)

### PR DESCRIPTION
**Pull request by tuannkhoi**
Closes: #90 

[testcase_mapreduce.md](https://github.com/nebula-labs/rust-by-example-vn/blob/main/src/std_misc/threads/testcase_mapreduce.md)
[arg.md](https://github.com/nebula-labs/rust-by-example-vn/blob/main/src/std_misc/arg.md)

### CODE REVIEW

File: src/std_misc/threads/testcase_mapreduce.md [hunk 1]
```diff
@@ -1 +1,144 @@
...
+### Gán giá trị
+Không nên để số lượng luồng phụ thuộc vào dữ liệu được nhập từ người dùng.
+Nếu người dùng quyết định nhập nhiều dấu cách, liệu chúng ta có muốn tạo ra 2,000 luồng không?
+Ta nên sửa đổi chương trình sao cho dữ liệu luôn được chia thành một số lượng nhỏ các phần, được xác định
+bởi một hằng số tĩnh ở đầu chương trình.
+
+### Xem thêm:
+* [Threads][thread]
+* [vectors][vectors] và [iterators][iterators]
+* [closures][closures], [move][move] semantics và [`move` closures][move_closure]
```
Comment (tuannkhoi) [inline src/std_misc/threads/testcase_mapreduce.md:129]: help me translate `semantics`

File: src/std_misc/arg.md [hunk 2]
```diff
@@ -1 +1,36 @@
...
+
+```shell
+$ ./args 1 2 3
+Đường dẫn của tôi là ./args.
+Tôi nhận được 3 đối số: ["1", "2", "3"].
+```
+
+## Crates
+
+Một cách khác là sử dụng các crate để cung cấp các chức năng phụ trợ hơn khi tạo ứng dụng dòng lệnh.
```
Comment (HongThaiPham) [inline src/std_misc/arg.md]:
```suggestion
Một cách khác là sử dụng các crate để cung cấp các chức năng bổ sung khi tạo ứng dụng dòng lệnh.
```

File: src/std_misc/arg.md [hunk 3]
```diff
@@ -1 +1,36 @@
...
+[Rust Cookbook] cho thấy những cách làm tốt nhất về cách sử dụng một trong những crate
```
Comment (HongThaiPham) [inline src/std_misc/arg.md]:
```suggestion
[Rust Cookbook] trình bày các quy tắc tốt nhất về cách sử dụng một trong những crate
```

File: src/std_misc/arg.md [hunk 4]
```diff
@@ -1 +1,36 @@
...
+đối số dòng lệnh phổ biến, `clap`.
```
Comment (HongThaiPham) [inline src/std_misc/arg.md]:
```suggestion
đối số dòng lệnh phổ biến nhất, là `clap`.
```

File: src/std_misc/threads/testcase_mapreduce.md [hunk 5]
```diff
@@ -1 +1,144 @@
 # Testcase: map-reduce
+
+Rust làm cho việc xử lý dữ liệu đa luồng trở nên rất dễ dàng mà không gặp phải
```
Comment (HongThaiPham) [inline src/std_misc/threads/testcase_mapreduce.md]:
```suggestion
Rust khiến cho việc xử lý dữ liệu đa luồng trở nên rất dễ dàng mà không đối mặt với
```

File: src/std_misc/threads/testcase_mapreduce.md [hunk 6]
```diff
@@ -1 +1,144 @@
...
+nhiều vấn đề thường gặp.
```
Comment (HongThaiPham) [inline src/std_misc/threads/testcase_mapreduce.md]:
```suggestion
các rắc rối thường gặp khi xử lý so với các ngôn ngữ khác.
```

File: src/std_misc/threads/testcase_mapreduce.md [hunk 7]
```diff
@@ -1 +1,144 @@
...
+
+Thư viện tiêu chuẩn cung cấp các nguyên tắc luồng tuyệt. Kết hợp với khái niệm Sở hữu và quy tắc
```
Comment (HongThaiPham) [inline src/std_misc/threads/testcase_mapreduce.md]:
```suggestion
Thư viện tiêu chuẩn cung cấp các nguyên mẫu đa luồng tuyệt vời sẵn có. Kết hợp với khái niệm Sở hữu và quy tắc
```

File: src/std_misc/threads/testcase_mapreduce.md [hunk 8]
```diff
@@ -1 +1,144 @@
...
+     ************************************************************************/
+
+    // Tách dữ liệu của chúng ta thành các phân đoạn cho từng phần tính toán
+    // Mỗi phân đoạn sẽ là một tham chiếu (&str) đến dữ liệu thực tế
+    let chunked_data = data.split_whitespace();
+
+    // Chạy qua các phân đoạn dữ liệu.
+    // .enumerate() thêm chỉ mục (index) của vòng lặp hiện tại vào bất cứ điều gì được chạy qua
+    // bộ đôi kết quả "(chỉ mục, phần tử)" sau đó được tự động
+    // "giải cấu trúc" thành hai biến, "i" và "data_segment" với một
```
Comment (HongThaiPham) [inline src/std_misc/threads/testcase_mapreduce.md]:
```suggestion
    // "destructured" thành hai biến, "i" và "data_segment" với một
```

File: src/std_misc/threads/testcase_mapreduce.md [hunk 9]
```diff
@@ -1 +1,144 @@
...
+    // "gán giải cấu trúc"
```
Comment (HongThaiPham) [inline src/std_misc/threads/testcase_mapreduce.md]:
```suggestion
    // "destructuring assignment"
```

File: src/std_misc/threads/testcase_mapreduce.md [hunk 10]
```diff
@@ -1 +1,144 @@
...
+        // * không có đối số ('||')
+        // * lấy sở hữu các biến bị chụp ('move') và
+        // * trả về một số nguyên 32-bit không dấu ('-> u32')
+        //
+        // Rust đủ thông minh để có thể suy luận được '-> u32' 
+        // từ chính closure nên ta có thể bỏ nó đi.
+        //
+        // TODO: thử xóa 'move' và xem điều gì sẽ xảy ra
+        children.push(thread::spawn(move || -> u32 {
+            // Tính tổng giữa kết quả tạm thời của đoạn dữ liệu này:
```
Comment (HongThaiPham) [inline src/std_misc/threads/testcase_mapreduce.md]:
```suggestion
            // Tính tổng trung gian của đoạn dữ liệu này:
```

File: src/std_misc/threads/testcase_mapreduce.md [hunk 11]
```diff
@@ -1 +1,144 @@
...
+            let result = data_segment
+                        // chạy qua các ký tự trong đoạn dữ liệu..
```
Comment (HongThaiPham) [inline src/std_misc/threads/testcase_mapreduce.md]:
```suggestion
                        // lặp qua các ký tự trong đoạn dữ liệu..
```

File: src/std_misc/threads/testcase_mapreduce.md [hunk 12]
```diff
@@ -1 +1,144 @@
...
+                        .chars()
+                        // lặp lại các ký tự trong đoạn dữ liệu..
```
Comment (HongThaiPham) [inline src/std_misc/threads/testcase_mapreduce.md]:
```suggestion
                        // chuyển ký tự sang kiểu số
```

File: src/std_misc/threads/testcase_mapreduce.md [hunk 13]
```diff
@@ -1 +1,144 @@
...
+                        .map(|c| c.to_digit(10).expect("nên là chữ số"))
+                        //  .. và tính tổng
```
Comment (HongThaiPham) [inline src/std_misc/threads/testcase_mapreduce.md]:
```suggestion
                        //  .. và tính tổng các số
```

File: src/std_misc/threads/testcase_mapreduce.md [hunk 14]
```diff
@@ -1 +1,144 @@
...
+                        .sum();
+
+            // println! khóa stdout, để không có tình trạng giao thoa văn bản xảy ra
```
Comment (HongThaiPham) [inline src/std_misc/threads/testcase_mapreduce.md]:
```suggestion
            // println! khóa stdout, để không có tình trạng văn bản xen kẽ nhau xảy ra
```

File: src/std_misc/threads/testcase_mapreduce.md [hunk 15]
```diff
@@ -1 +1,144 @@
...
+     * Giai đoạn "Reduce"
+     *
+     * Tổng hợp kết quả trung gian và kết hợp chúng thành kết quả cuối cùng
+     ************************************************************************/
+
+    // Kết hợp kết quả trung gian của từng thread thành một tổng kết quả cuối cùng.
+    //
+    // chúng ta sử dụng "turbofish" ::<> để cung cấp cho sum() một gợi ý kiểu dữ liệu.
+    //
+    // TODO: hãy thử không sử dụng turbofish, thay vì đó chỉ rõ kiểu của final_result
```
Comment (HongThaiPham) [inline src/std_misc/threads/testcase_mapreduce.md]:
```suggestion
    // TODO: hãy thử không sử dụng turbofish, thay vào đó chỉ rõ kiểu của final_result
```

**Review by HongThaiPham — APPROVED** [review]

**Pull request merged**